File: backend.py
import random
from flask import Flask, request, jsonify, render_template
from flask_cors import CORS, cross_origin

app = Flask(__name__, static_folder="dist/assets", template_folder='dist')
cors = CORS(app)
app.config['CORS_HEADERS'] = 'Content-Type'



def LoadValidWords():
    result = []
    with open("valid_words.txt") as f:
        for line in f.readlines():
            result.append(line.strip().upper())
    return result


valid_words = LoadValidWords()


class LetterStatus:
    NOT_SET = "NOT_SET"
    CORRECT = "CORRECT"
    PRESENT = "PRESENT"
    ABSENT = "ABSENT"

class LetterInfo:
    def __init__(self, letter, status) -> None:
        self.letter = letter
        self.status = status


def findFirstAndRemove(remaining_letters : list, char : str):
    is_present = (char in remaining_letters)
    if is_present:
        remaining_letters.remove(char)
    return is_present
        

def to_letters(input : str):
    result = [LetterInfo(input[i],LetterStatus.ABSENT) for i in range(5)]
    remaining_letters = [char for char in answer]
    remaining_idxs = []

    # Check for correct ones
    for i in range(5):
        if input[i] == answer[i]:
            result[i].status =  LetterStatus.CORRECT
            remaining_letters.remove(input[i])
        else:
            remaining_idxs.append(i)
    
    # Check for existing ones
    for i in remaining_idxs:
        if input[i] in remaining_letters:
            result[i].status = LetterStatus.PRESENT
            remaining_letters.remove(input[i])

    return [letter.__dict__ for letter in result]

@app.route('/reset', methods=['GET'])
@cross_origin()
def reset():
    global attempt
    attempt = 0
    return jsonify({'all': 'good'})


@app.route('/check', methods=['POST'])
@cross_origin()
def reverse_text():
    global attempt
    global valid_words
    data = request.json
    if 'text' in data:
        try:
            word = data["text"].upper()
            if word not in valid_words:
                return jsonify({'error': 'Invalid word'}), 422
            if len(word) != 5:
                return jsonify({'error': 'Word must have 5 characters'}), 422

            response = to_letters(word)
            attempt += 1                                
            return jsonify({'result': response, 'attempt' : attempt})
        except Exception as e:
            return jsonify({'error': e.args[0]}), 500

    else:
        return jsonify({'error': 'Text not provided'}), 400


@app.route('/')
def home():
    return render_template('index.html')


if __name__ == '__main__':
    attempt = 0
    answer = valid_words[random.randint(0, len(valid_words) - 1)]
    app.run(debug=True)

File: src/Backend.ts
import LetterProps from "./LetterProps";
import LetterStatus from "./LetterStatus";
import valid_words from "./valid_words";



function removeFirstOccurrence<T>(arr: T[], element: T): void {
    // Find the index of the first occurrence of the element
    const index = arr.indexOf(element);

    // If the element is found in the array
    if (index !== -1) {
        // Use splice to remove the element at the found index
        arr.splice(index, 1);
    }

}



//constructor() {
//    this.valid_words = valid_words;
//    const n: number = Math.floor(Math.random() * valid_words.length);
//    this.answer = valid_words[n];
//    this.validateWord = this.validateWord.bind(this);
//    console.log(this.answer);
//}


function validateWord(answer: string, guess: string): LetterProps[] {
    let result: LetterProps[] = [];
    let remaining_letters: string[] = [];
    let remaining_indexes: number[] = [];

    guess = guess.toUpperCase();

    if (guess.length != 5)
        return [];
    if (!valid_words.includes(guess))
        return [];


    for (const letter of guess) {
        result.push({ letter: letter, status: LetterStatus.ABSENT });
    }
    for (const letter of answer) {
        remaining_letters.push(letter);
    }


    // Check for correct ones
    for (let i: number = 0; i < 5; i++) {
        if (guess[i] === answer[i]) {
            result[i].status = LetterStatus.CORRECT;
            removeFirstOccurrence(remaining_letters, result[i].letter);
        } else {
            remaining_indexes.push(i);
        }
    }

    // Check for existing ones
    for (const i of remaining_indexes) {

        if (remaining_letters.includes(guess[i])) {

            result[i].status = LetterStatus.PRESENT
            removeFirstOccurrence(remaining_letters, guess[i]);
        }

    }

    return result;
}


export default validateWord;